### PR TITLE
done

### PR DESCRIPTION
**Pull request by Ppllupold**

### CODE REVIEW

File: src/main/java/practice/CandidateValidator.java [hunk 1]
```diff
@@ -1,5 +1,24 @@
...
+        return candidate.getAge() >= START_AGE_FOR_VOTING
+                && candidate.isAllowedToVote()
+                && candidate.getNationality().equals("Ukrainian")
+                && validatePeriodInUkr(candidate.getPeriodsInUkr());
+    }
+
+    private static boolean validatePeriodInUkr(String period) {
+        String[] split = period.split("-");
+        return Integer.parseInt(split[split.length - 1])
+                - Integer.parseInt(split[0]) >= MIN_ALLOWED_PERIOD_IN_UKR;
```
Comment (MentorMA) [inline src/main/java/practice/CandidateValidator.java:21]: magic numbers

File: src/main/java/practice/CandidateValidator.java [hunk 2]
```diff
@@ -1,5 +1,24 @@
...
+
+public class CandidateValidator implements Predicate<Candidate> {
+    private static final int START_AGE_FOR_VOTING = 35;
+    private static final int MIN_ALLOWED_PERIOD_IN_UKR = 10;
+
+    @Override
+    public boolean test(Candidate candidate) {
+        return candidate.getAge() >= START_AGE_FOR_VOTING
+                && candidate.isAllowedToVote()
+                && candidate.getNationality().equals("Ukrainian")
```
Comment (MentorMA) [inline src/main/java/practice/CandidateValidator.java:14]: String should be constant

File: src/main/java/practice/StreamPractice.java [hunk 3]
```diff
@@ -23,7 +34,15 @@ public int findMinEvenNumber(List<String> numbers) {
...
-        return 0D;
+        List<Integer> adjustedNumbers = IntStream.range(0, numbers.size())
+                .mapToObj(i -> (i % 2 != 0) ? numbers.get(i) - 1 : numbers.get(i))
+                .collect(Collectors.toList());
+
+        return adjustedNumbers.stream()
+                .filter(num -> num % 2 != 0)
+                .mapToInt(Integer::intValue)
+                .average()
+                .orElseThrow();
```
Comment (MentorMA) [inline src/main/java/practice/StreamPractice.java:45]: Use one stream for this method, just combine

File: src/main/java/practice/StreamPractice.java [hunk 4]
```diff
@@ -35,7 +54,10 @@ public Double getOddNumsAverage(List<Integer> numbers) {
      * Example: select men who can be recruited to army (from 18 to 27 years old inclusively).
      */
     public List<Person> selectMenByAge(List<Person> peopleList, int fromAge, int toAge) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(p -> p.getSex().equals(Person.Sex.MAN))
+                .filter(p -> p.getAge() >= fromAge && p.getAge() <= toAge)
```
Comment (MentorMA) [inline src/main/java/practice/StreamPractice.java:59]: you can use one filter method

File: src/main/java/practice/StreamPractice.java [hunk 5]
```diff
@@ -59,7 +85,14 @@ public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
...
      */
     public List<String> getCatsNames(List<Person> peopleList, int femaleAge) {
-        return Collections.emptyList();
+
+        return peopleList.stream()
+                .filter(person -> person.getAge() >= femaleAge
+                        && person.getSex().equals(Person.Sex.WOMAN))
+                .map(Person::getCats)
+                .flatMap(Collection::stream)
+                .map(Cat::getName)
```
Comment (MentorMA) [inline src/main/java/practice/StreamPractice.java:94]: too many operations map

File: src/main/java/practice/StreamPractice.java [hunk 6]
```diff
@@ -75,6 +108,13 @@ public List<String> getCatsNames(List<Person> peopleList, int femaleAge) {
      * parametrized with Candidate in CandidateValidator.
      */
     public List<String> validateCandidates(List<Candidate> candidates) {
-        return Collections.emptyList();
+        if (candidates.isEmpty()) {
+            return Collections.emptyList();
+        }
```
Comment (MentorMA) [inline src/main/java/practice/StreamPractice.java:113]: this check not mandatory

**Review by MentorMA — CHANGES_REQUESTED** [review]

**Review by Elena-Bruyako — APPROVED** [review]
Approve, but there is spelling mistake in country constant, also you can use toList(); instead of .collect(Collectors.toList());

**Pull request closed without merging**